### PR TITLE
Parallelize the LSS pipeline

### PR DESCRIPTION
**Pull request by ljgray**
Ongoing PR to parallelize and speed up the LSS tasks. 

Most tasks are already quite fast - the only bottleneck is `signal.corrfunc.corr_to_clarray`, with the main bottleneck there being the calls to `util.cubicspline.SinhInterpolator`. The interpolator is already openmp parallel, so adding MPI parallelization requires using multiple nodes to get any speed-up.

### CODE REVIEW

File: cora/signal/lss.py [hunk 1]
```diff
@@ -648,12 +642,25 @@ def process(self, initial_field: InitialLSS, biased_field: BiasedLSS) -> BiasedL
...
         self._validate_fields(initial_field, biased_field)
         c, nside, _, chi, za = self._get_props(biased_field)
 
         # Take the gradient of the Lagrangian potential to get the
-        # displacements
-        vpsi = lssutil.gradient(initial_field.phi[:], chi)
+        # displacements. Need the full array on each rank because
+        # we want the gradient along both axes. This is a bit of an
+        # awkward way to do this, but we have to do it this way to
+        # avoid ballooing the memory
```
Comment (sjforeman) [inline cora/signal/lss.py:657]:
```suggestion
        # avoid ballooning the memory
```

File: cora/signal/lssutil.py [hunk 2]
```diff
@@ -233,6 +233,10 @@ def gradient(maps: np.ndarray, x: np.ndarray) -> np.ndarray:
         An 2D array of maps. First axis is the x coord, second is Healpix pixels.
     x
         The coordinate along the first axis. This axis need not be uniform.
+    grad0
+        If True, compute the gradient along axis 0. Otherwise, output `grad[0]`
+        will not be computed. This is usefull if `maps` is distributed.
```
Comment (sjforeman) [inline cora/signal/lssutil.py:238]:
```suggestion
        will not be computed. This is useful if `maps` is distributed.
```

File: cora/signal/lss.py [hunk 3]
```diff
@@ -311,23 +305,24 @@ def process(self) -> InitialLSS:
...
+
+        # Create an extended covariance matrix capturing the cross correlations between
+        # the fields
+        cla = mpiarray.zeros((lmax + 1, 2 * nz, 2 * nz), axis=0)
 
         self.log.debug("Generating C_l(x, x')")
         cla0 = corrfunc.corr_to_clarray(corr0, lmax, xa, xromb=self.xromb, q=self.leg_q)
-        cla2 = corrfunc.corr_to_clarray(corr2, lmax, xa, xromb=self.xromb, q=self.leg_q)
-        cla4 = corrfunc.corr_to_clarray(corr4, lmax, xa, xromb=self.xromb, q=self.leg_q)
+        cla.local_array[:, nz:, nz:] = cla0
```
Comment (sjforeman) [inline cora/signal/lss.py:316]: Do we need `cla0.local_array` here, or is that redundant?

**Review by sjforeman — APPROVED** [review]
Looks great! Just 2 docstring typos and one question

File: cora/signal/lss.py (hunk 3 above)
Comment (ljgray) [reply cora/signal/lss.py:316]: It shouldn't matter. I actually don't think we even need `cla.local_array`, it should be happy to write one MPIArray into another. I'll do a quick test and change to be consistent

**Pull request merged**